src/sysinfo

```diff
@@ -1175,6 +1177,55 @@ fi
 END
 }
 
+function valid_uuid(){
+        uuid -d "$@" &>/dev/null
+}
+
+# Hardware can be fickle. We have seen systems with a misbehaving
+# BMC return the wrong UUID to `smbios` on later boots.
+# We want to record the node UUID on the zones pool and keep reusing
+# it on subsequent boots.
+# Sources of UUID:
+# 1. `smbios`
+# 2. override_uuid bootparam
+# 3. org.smartos:server_uuid on $Zpool
+# We prefer 3 over 1
+# We prefer 2 over 3 and update 3 if they disagree.
+function get_or_store_uuid()
+{
+    [[ -z "$Zpool" ]] && return
+
+    local key="org.smartos:server_uuid"
+    local dataset="${Zpool}/var"
+    local uuid_source=$(zfs get -Hp -o source ${key} ${dataset})
+    if [[ "$uuid_source" == "local" ]]; then
+        # Read what was stored on the pool.
+        local uuid_value=$(zfs get -Hp -o value ${key} ${dataset})
+        # If we received a valid override, update the pool if needed.
+        if [[ -n ${OVERRIDE_UUID} && ${OVERRIDE_UUID} != ${uuid_value} ]]; then
+            valid_uuid ${OVERRIDE_UUID} && \
+                zfs set ${key}=${OVERRIDE_UUID} ${dataset}
+            # Explicitly read it back in.
+            uuid_value=$(zfs get -Hp -o value ${key} ${dataset})
+        fi
+        if valid_uuid ${uuid_value}; then
+            # Pool contains a valid UUID. Use it.
+            UUID=${uuid_value}
```

Comment:
Upon further review, do we want to complain somewhere if the existing `UUID` is not the same as `uuid_value` for diagnostic purposes? I don't think there's a wrong answer here beyond, "no idea." Given how frequently this gets called, maybe not?

Reply:
You've made a good case for no-need.  Resolving.